src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs

```diff
@@ -27,35 +28,45 @@ public abstract class HttpClientHandlerTest_Http2 : HttpClientHandlerTestBase
 
         public HttpClientHandlerTest_Http2(ITestOutputHelper output) : base(output) { }
 
-        private async Task AssertProtocolErrorAsync(Task task, ProtocolErrors errorCode)
+        private async Task AssertProtocolErrorAsync(Task task, ProtocolErrors errorCode, bool allowConnectionRefused = false)
         {
             Exception e = await Assert.ThrowsAsync<HttpRequestException>(() => task);
             if (UseSocketsHttpHandler)
             {
                 string text = e.ToString();
+                if (allowConnectionRefused && (
+                    text.Contains("No connection could be made because the target machine actively refused it.") ||
```

Comment:
Checking exception text like this is brittle. And if we run on another language machine the text will be in that language so this check will fail.

Instead we should be checking the inner exception (most likely a SocketException) for error codes that match these strings such as:

* WSAECONNREFUSED
* WSAECONNRESET
etc.

Reply:
@davidsh good point, I'll see if I can improve it somehow

Reply:
```csharp
        private static bool IsExceptionConnectionRefused(Exception e)
        {
            while (e != null)
            {
                if (e is Sockets.SocketException sockException)
                {
                    if (sockException.SocketErrorCode == Sockets.SocketError.ConnectionRefused)
                    {
                        return true;
                    }
                }

                e = e.InnerException;
            }

            return false;
        }
```

Something like this somewhat works so far but considering this is an edge case I'll need to leave it running for a bit to check if this covers all cases. I'll update PR while I'm testing